extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js

```diff
@@ -264,6 +266,53 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
     if (this.ampAnalyticsConfig_) {
       insertAnalyticsElement(this.element, this.ampAnalyticsConfig_, true);
     }
+
+    analyticsForDoc(this.getAmpDoc(), true).then(analytics => {
+      const vis = analytics.getAnalyticsRoot(this.element).getVisibilityManager();
+      // Can be any promise or `null`.
+      const readyPromise = Promise.race([
+        this.signals().whenSignal(CommonSignals.INI_LOAD),
+        this.signals().whenSignal(CommonSignals.LOAD_END),
+      ]);
+      // Element must be an AMP element at this time.
+      debugger;
+      // 50% vis w/o ini load
+      vis.listenElement(this.element, {visiblePercentageMin: 50}, null,
+                        event => {
+                          console.log(event);
+                        });
+      // 50% vis w ini load
+      vis.listenElement(this.element,
+                        {visiblePercentageMin: 50, waitFor: 'ini-load'},
+                        this.signals().whenSignal(CommonSignals.INI_LOAD),
```

Comment:
This should be `readyPromise`. The trick is that you always need to race ini-promise and load-end, otherwise ini-promise may never show up.